XD-3534 Add HDFS Dataset sink module

trisberg opened the pull request:
- Ported existing Spring XD hdfs-dataset sink module

This resolves #105.

trisberg committed:
XD-3534 Add HDFS Dataset sink module

- Ported existing Spring XD hdfs-dataset sink module

artembilan commented on hdfs-dataset-sink/pom.xml:
```diff
@@ -0,0 +1,66 @@
+	</parent>
+
+	<properties>
+		<start-class>org.springframework.cloud.stream.module.hdfs.dataset.DatasetSinkApplication</start-class>
```
Looks like should be this one `org.springframework.cloud.stream.module.dataset.sink`...

trisberg replied:
Indeed, forgot to change this after changing the package name. Fixing.

artembilan commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSink.java:
```diff
@@ -0,0 +1,64 @@
+			}
+			datasetOperations.write((Collection<?>) message.getPayload());
+		} else {
+			logger.warn("Expected a collection of POJOs but received " + message.getPayload().getClass().getName());
```
Not sure that it is a big issue just have a single POJO and store it as is without any warn...
`else` on new line, BTW

trisberg replied:
It indicates something wrong in the configuration and it would be bad since writing each object individually would create lots of small files in HDFS.

artembilan commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkConfiguration.java:
```diff
@@ -0,0 +1,190 @@
+		AggregatorFactoryBean aggregatorFactoryBean = new AggregatorFactoryBean();
+		aggregatorFactoryBean.setCorrelationStrategy(
+				new ExpressionEvaluatingCorrelationStrategy("payload.getClass().name"));
+		String releaseStrategyExpression = "size() == " + properties.getBatchSize();
```
Looks like `MessageCountReleaseStrategy` would be better for this kind of aggregator.

trisberg replied:
Just porting what we have in XD, will give `MessageCountReleaseStrategy` a try.

artembilan commented on hdfs-dataset-sink/src/test/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkConfigurationTests.java:
```diff
@@ -0,0 +1,87 @@
+
+	@After
+	public void checkFilesClosedOK() throws IOException, InterruptedException {
+		applicationContext.close();
```
What is the reason to `close()` ctx manually if you are based on the Spring Test Framework?
Especially it will be very dangerous when you add one more `@Test`  to this class

trisberg replied:
Seems to be needed for at least one of the tests, I'll remove the ones not needed.

artembilan commented on hdfs-dataset-sink/src/test/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkIntegrationTests.java:
```diff
@@ -0,0 +1,221 @@
+		public void checkFilesClosedOK() throws IOException {
+			applicationContext.close();
+			File testOutput = new File(testDir);
+			assertTrue("Dataset path created", testOutput.exists());
```
That looks weird to `assert` in the `@After`...

trisberg replied:
Yeah, not necessary for these tests, will remove.

artembilan commented on hdfs-dataset-sink/src/test/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkIntegrationTests.java:
```diff
@@ -0,0 +1,221 @@
+			"idleTimeout=2000"})
+	public static class PartitionedDatasetTests extends DatasetSinkIntegrationTests {
+
+		Long systemTime = 1446570266690L; // Date is 2015/11/03
```
What is the purpose of this?
Why `System.currentTimeMillis()` isn't enough if that?

trisberg replied:
Just to have some control, want the messages to always fall within the same year.

artembilan commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSink.java:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	@ServiceActivator(inputChannel="toSink")
+	public void datasetSink(Message<?> message) {
```
I'd avoid this class in favor of simple `@Bean` for an inline `MessageHandler` impl.
With this `@ServiceActivator(inputChannel="toSink")`, of course.
From other side the same can be utilized with the SI Java DSL. Especially that horror around the `AggregatorFactoryBean`...

artembilan commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkConfiguration.java:
```diff
@@ -0,0 +1,190 @@
+import org.springframework.util.StringUtils;
+
+import java.util.Arrays;
+import java.util.List;
```
What import rules do you follow?
https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Code-Style:

> The import statements are structured as follow:

```
import java.*
import javax.*
blank line
import all other imports
blank line
import org.springframework.*
blank line
import static all other imports
```

> Also, static imports should not be used in production code. They should be used in test code, especially for things like org.junit.Assert.

artembilan commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkConfiguration.java:
```diff
@@ -0,0 +1,190 @@
+			catch (SpelParseException spe) {
+				if (!expr.trim().endsWith(")")) {
+					throw new StoreException("Invalid partitioning expression '" + expr
+							+ "' -  did you forget the closing parenthesis?", spe);
```
I'm not sure that it is a good idea to say that user is fool so openly. :smile:

ericbottard commented on hdfs-dataset-sink/pom.xml:
```diff
@@ -0,0 +1,66 @@
+			<groupId>org.springframework.data</groupId>
+			<artifactId>spring-data-hadoop-boot</artifactId>
+			<version>${spring-data-hadoop.version}</version>
+			<scope>provided</scope>
```
I think provided should be removed here

trisberg replied:
We need that if we want to switch in another hadoop distro using --includes

ericbottard replied:
Hmm, not sure I understand exactly how distro switching would work in your opinion.
Do you expect to create BOMs that transitively include `spring-data-hadoop-boot` and `spring-data-hadoop-store` for distros? If so, and several distros depend on different versions of the above, then the code for this module may fail at runtime anyway (given we would provide a different version of say, `org.springframework.data.hadoop.store.dataset.DatasetDefinition` from what was used at compile time)

ericbottard commented on hdfs-dataset-sink/pom.xml:
```diff
@@ -0,0 +1,66 @@
+			<groupId>org.springframework.data</groupId>
+			<artifactId>spring-data-hadoop-store</artifactId>
+			<version>${spring-data-hadoop.version}</version>
+			<scope>provided</scope>
```
ditto

trisberg replied:
ditto

ericbottard commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkApplication.java:
```diff
@@ -0,0 +1,34 @@
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+/**
+ * Spring Boot App for starting the {@link DatasetSink} module
```
Missing dot at the end

ericbottard commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSink.java:
```diff
@@ -0,0 +1,64 @@
+			Collection<?> payloads = (Collection<?>) payload;
+			if (logger.isDebugEnabled()) {
+				logger.debug("Writing a collection of " + payloads.size() +
+						" POJOs of type " + payloads.toArray()[0].getClass().getName());
```
We may want to avoid `toArray()` just for logging

ericbottard replied:
And the collection could be empty...

trisberg replied:
changing the logging here, don't need to be this detailed

ericbottard commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkConfiguration.java:
```diff
@@ -0,0 +1,190 @@
+				Expression e = parser.parseExpression(expr);
+				psb = e.getValue(ctx, PartitionStrategy.Builder.class);
+			}
+			catch (SpelParseException spe) {
```
This would rather be caught at injection time by having an `Expression[]` filed + getter/setter

trisberg replied:
trying to keep the functionality from XD where partition expressions are separated by '/' character

ericbottard commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkProperties.java:
```diff
@@ -0,0 +1,162 @@
+	private int writerCacheSize = -1;
+
+	/**
+	 * Compression type name (snappy, deflate, bzip2 (avro only) or uncompressed)
```
Can this be made into an enum, especially if we document those as the only acceptable values.

ericbottard commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkProperties.java:
```diff
@@ -0,0 +1,162 @@
+	private String partitionPath = null;
+
+	/**
+	 * The size of the cache to be used for partition writers (10 if omitted).
```
If 10 is the default, why not initialize at 10 (and remove doc about the default)?

trisberg replied:
if it's not specified we don't want to set this since that adds a runtime system property just for this rarely used setting, so we need to keep track of when this property is not set - all this gets handled by Kite SDK where the current default they use is 10

ericbottard replied:
My point is, especially if Kite is responsible for the default, text documentation of defaults can become stale. actual values that are always set don't lie

ericbottard commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkProperties.java:
```diff
@@ -0,0 +1,162 @@
+	/**
+	 * The format to use, valid options are avro and parquet.
+	 */
+	private String format = "avro";
```
enum?

trisberg replied:
didn't have one in XD so out of scope ;)

ericbottard replied:
myyyeaahhhh, but we have better DSL code completion with enums now :)
Let's open a ticket to track those

ericbottard commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkProperties.java:
```diff
@@ -0,0 +1,162 @@
+	/**
+	 * The base directory path where the files will be written in the Hadoop FileSystem.
+	 */
+	private String directory = "/tmp/hdfs-dataset-sink";
```
Should this depend on the "stream name"? ie `@Value("${spring.application.name:foo}")`

ericbottard replied:
Ah, I see the following property is built that way. nvm

ericbottard commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkProperties.java:
```diff
@@ -0,0 +1,162 @@
+	private String directory = "/tmp/hdfs-dataset-sink";
+
+	/**
+	 * The sub-directory under the basePath where files will be written.
```
this refers to `basePath` but is called `directory` above, correct?

trisberg replied:
yes, will clarify the javadoc

ericbottard commented on hdfs-dataset-sink/src/test/java/org/springframework/cloud/stream/module/dataset/domain/TestPojo.java:
```diff
@@ -0,0 +1,53 @@
+package org.springframework.cloud.stream.module.dataset.domain;
+
+/**
+ * Test POJO for writing partitioned dataset
```
Missing dot (and `@author`)

ericbottard commented on hdfs-dataset-sink/src/test/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkConfigurationTests.java:
```diff
@@ -0,0 +1,87 @@
+@SpringApplicationConfiguration(classes = DatasetSinkApplication.class)
+@WebIntegrationTest({"server.port:0",
+		"spring.hadoop.fsUri=file:///",
+		"directory=${java.io.tmpdir}/dataset",
```
Javadoc on the Properties class says this is a Path in Hadoop

trisberg replied:
it is, but with the fsUri set to file:/// it actually writes to the local file system through hdfs, so using a tmp directory made sense

ericbottard replied:
I see.
Wouldn't it be better to move the ${java.io.tmpdir} part to
`spring.hadoop.fsUri`
though (`spring.hadoop.fsUri=file://${java.io.tmpdir}`) and drop it from
`directory` ?

On Wed, Nov 4, 2015 at 9:25 PM, Thomas Risberg notifications@github.com
wrote:

> In
> hdfs-dataset-sink/src/test/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkConfigurationTests.java
> https://github.com/spring-cloud/spring-cloud-stream-modules/pull/58#discussion_r43934442
> :
> 
> > +import org.springframework.test.annotation.DirtiesContext;
> > +import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
> > +
> > +import java.io.File;
> > +import java.io.IOException;
> > +
> > +import static org.junit.Assert.assertTrue;
> > +
> > +/**
> > - \* @author Thomas Risberg
> > - */
> >   +@RunWith(SpringJUnit4ClassRunner.class)
> >   +@SpringApplicationConfiguration(classes = DatasetSinkApplication.class)
> >   +@WebIntegrationTest({"server.port:0",
> > -       "spring.hadoop.fsUri=file:///",
> > -       "directory=${java.io.tmpdir}/dataset",
> 
> it is, but with the fsUri set to file:/// it actually writes to the local
> file system through hdfs, so using a tmp directory made sense
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/spring-cloud/spring-cloud-stream-modules/pull/58/files#r43934442
> .

ericbottard commented on hdfs-dataset-sink/src/test/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSinkConfigurationTests.java:
```diff
@@ -0,0 +1,87 @@
+		"batchSize=2",
+		"idleTimeout=2000"})
+@DirtiesContext
+public class DatasetSinkConfigurationTests {
```
This does not really seem to test _configuration_.
Moreover, the test about file being ok when context closes is also in another class, unless I miss something

trisberg replied:
yeah, removing this test since other tests cover this

ericbottard commented on hdfs-dataset-sink/src/test/resources/logback.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This should be removed

trisberg replied:
why, without it we get a ton of debug logging when the test class is loaded

jvalkeal commented on hdfs-dataset-sink/src/main/java/org/springframework/cloud/stream/module/dataset/sink/DatasetSink.java:
```diff
@@ -0,0 +1,64 @@
+		Object payload = message.getPayload();
+		if (payload instanceof Collection<?>) {
+			Collection<?> payloads = (Collection<?>) payload;
+			if (logger.isDebugEnabled()) {
```
Others have nitpicked to me that with slf4j `logger.isDebugEnabled()` is not needed anymore if you use parameterized messages within a log statement. Thought if you don't change debug call then this is needed. Nitpicking... :)

trisberg replied:
ok, changing the logging

trisberg committed:
XD-3534 Add HDFS Dataset sink module

- Addressing review comments/suggestions

jvalkeal commented:
I have some issues with `idleTimeout` and `fsUri`:

Running local `hdfs://localhost:8020` and I simply build the module locally and registered module with:
```
module register --coordinates org.springframework.cloud.stream.module:hdfs-dataset-sink:jar:exec:1.0.0.BUILD-SNAPSHOT --name hdfs-dataset-sink --type sink
```

`idleTimeout` defaults to `0` so it seems that every single write creates a new dataset so I'm getting new avro file every 2 seconds. Being honest I don't remember if XD dataset sink behaves like this. I'd expect `idleTimeout=0` to disable flushing and thus simply falling back to `batchSize` when new avro file is created.
```
stream create --name "ticktock" --definition "time --fixedDelay=2|hdfs-dataset" --deploy
```

```
drwxr-xr-x   - root supergroup          0 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data
drwxr-xr-x   - root supergroup          0 2015-11-05 07:43 /tmp/hdfs-dataset-sink/data/string
drwxr-xr-x   - root supergroup          0 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/.metadata
-rw-r--r--   3 root supergroup        173 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/.metadata/descriptor.properties
-rw-r--r--   3 root supergroup          8 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/.metadata/schema.avsc
drwxr-xr-x   - root supergroup          0 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/.metadata/schemas
-rw-r--r--   3 root supergroup          8 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/.metadata/schemas/1.avsc
-rw-r--r--   3 root supergroup        105 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/0142625d-86f4-4034-be0a-a223ca78fdba.avro
-rw-r--r--   3 root supergroup        105 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/040ce3e4-46c0-4602-a8a7-f8c2f48fb6ea.avro
-rw-r--r--   3 root supergroup        105 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/041535af-c176-438f-8744-ad53c15a9c00.avro
-rw-r--r--   3 root supergroup        116 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/1c38c57d-ac86-4140-819d-0f648f9dc84c.avro
-rw-r--r--   3 root supergroup        105 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/294c5fa0-ebb5-403d-96aa-2a76cf0f01a5.avro
-rw-r--r--   3 root supergroup        105 2015-11-05 07:42 /tmp/hdfs-dataset-sink/data/string/2c0dbc08-e98e-4286-bc37-67d67017e038.avro
```

Tried to use `20sec` `idleTimeout` but then it waits default `batchSize` `10000` so didn't get any writes. Expected and didn't want to wait batch to reach its size.
```
stream create --name "ticktock" --definition "time --fixedDelay=2|hdfs-dataset --idleTimeout=20000" --deploy
```

With `batchSize=10` I get 3 avro files per minute which is expected if `idleTimeout` is more that 2 sec.
```
stream create --name "ticktock" --definition "time --fixedDelay=2|hdfs-dataset --idleTimeout=20000 --batchSize=10" --deploy
```

I don't see `fsUri` used anywhere in `DatasetSinkConfiguration`. It seem to use auto-config for `Configuration` which defaults to `hdfs://localhost:8020`. So expected stream writes to fail if giving spoofed `fsUri`. Files were still written under custom location.
```
stream create --name "ticktock" --definition "time --fixedDelay=2|hdfs-dataset --fsUri=hdfs://localhost:12345 --directory=/tmp/hdfs-dataset-sink-custom" --deploy
```

```
root@neo:/usr/local/hadoops# hadoop/bin/hdfs dfs -ls /tmp
Found 2 items
drwxrwxrwx   - root supergroup          0 2015-09-02 08:36 /tmp/hadoop-yarn
drwxr-xr-x   - root supergroup          0 2015-11-05 08:11 /tmp/hdfs-dataset-sink-custom
root@neo:/usr/local/hadoops# hadoop/bin/hdfs dfs -ls -R /tmp/hdfs-dataset-sink-custom
drwxr-xr-x   - root supergroup          0 2015-11-05 08:11 /tmp/hdfs-dataset-sink-custom/data
drwxr-xr-x   - root supergroup          0 2015-11-05 08:11 /tmp/hdfs-dataset-sink-custom/data/string
drwxr-xr-x   - root supergroup          0 2015-11-05 08:11 /tmp/hdfs-dataset-sink-custom/data/string/.metadata
-rw-r--r--   3 root supergroup        180 2015-11-05 08:11 /tmp/hdfs-dataset-sink-custom/data/string/.metadata/descriptor.properties
-rw-r--r--   3 root supergroup          8 2015-11-05 08:11 /tmp/hdfs-dataset-sink-custom/data/string/.metadata/schema.avsc
drwxr-xr-x   - root supergroup          0 2015-11-05 08:11 /tmp/hdfs-dataset-sink-custom/data/string/.metadata/schemas
-rw-r--r--   3 root supergroup          8 2015-11-05 08:11 /tmp/hdfs-dataset-sink-custom/data/string/.metadata/schemas/1.avsc
-rw-r--r--   3 root supergroup        110 2015-11-05 08:11 /tmp/hdfs-dataset-sink-custom/data/string/0274cc54-572b-4809-b11f-ab8e18ceafa6.avro
-rw-r--r--   3 root supergroup        105 2015-11-05 08:11 /tmp/hdfs-dataset-sink-custom/data/string/033fc459-21ff-4763-84a0-2b48f8e27980.avro
```

trisberg committed:
XD-3534 Add HDFS Dataset sink module

- Addressing some deployment issues

trisberg committed:
XD-3534 Add HDFS Dataset sink module

- Fixing tests after we removed the autoconfigured FsShell

trisberg committed:
XD-3534 Add HDFS Dataset sink module

- Adding crreation of separate configuration for `--fsUri` option

jvalkeal commented:
Yes, I'm seeing same `java.io.IOException: Filesystem closed` error. Trying to figure out where it's coming from, in case it's not caused by shutdown hook.

jvalkeal commented:
Interesting, it seems that I don't get `Filesystem closed` error if I do:
```
stream create --name "ticktock" --definition "time|hdfs-dataset --batchSize=20" --deploy
```
and wait first file to be written.

If I do:
```
stream create --name "ticktock" --definition "time|hdfs-dataset" --deploy
```
there's no files written(if I don't wait full default batch size) and I get error.

It's worth to check if we get this error only if no data is flushed, thus no interaction with kite, thus there would not be any existing open hadoop FileSystem's.

jvalkeal commented:
damn, I take it back. This error doesn't happen every stream undeploy.

jvalkeal commented:
Right, it has to be `FileSystem` autoclose which is i.e. called from shutdown hook. I don't see errors if I disable autoclose with:
```
stream create --name "ticktock" --definition "time|hdfs-dataset --spring.hadoop.config.fs.automatic.close=false" --deploy
```

trisberg committed:
XD-3534 Add HDFS Dataset sink module

- Adding bean to control FileSystem close, rather than rely on shutdown hook registered by Hadoop

trisberg committed:
XD-3534 Add HDFS Dataset sink module

- Remove unused configuration

trisberg commented:
Rebased

jvalkeal commented:
squashed and merged per https://github.com/spring-cloud/spring-cloud-stream-modules/commit/3181e73a8e3e737844e9f249bcdd76043cf03eba
